helios-crtauth/src/main/java/com/spotify/helios/auth/crt/CrtAuthProvider.java

```diff
@@ -0,0 +1,127 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.Response.Status;
+
+import io.dropwizard.auth.Auth;
+import io.dropwizard.auth.AuthenticationException;
+
+/**
+ * This class sets up Jersey to be able to call our Authenticator instance whenever a constructor or
+ * method parameter is annotated with {@link Auth}.
+ */
+public class CrtAuthProvider implements InjectableProvider<Auth, Parameter> {
+
+  private final CrtTokenAuthenticator tokenAuthenticator;
+
+  public CrtAuthProvider(CrtTokenAuthenticator tokenAuthenticator) {
+    this.tokenAuthenticator = tokenAuthenticator;
+  }
+
+  @Override
+  public ComponentScope getScope() {
+    return ComponentScope.PerRequest;
+  }
+
+  @Override
+  public Injectable getInjectable(ComponentContext ic, Auth auth, Parameter parameter) {
+    return new CrtAuthInjectable(tokenAuthenticator, auth.required());
+  }
+
+  private static class CrtAuthInjectable extends AbstractHttpContextInjectable<HeliosUser> {
+
+    private static final Logger log = LoggerFactory.getLogger(CrtAuthInjectable.class);
+
+    private final CrtTokenAuthenticator authenticator;
+    private final boolean required;
+
+    public CrtAuthInjectable(CrtTokenAuthenticator authenticator, boolean required) {
+      this.authenticator = authenticator;
+      this.required = required;
+    }
+
+    /**
+     * The role of this method is to examine the HttpContext for the Authorization header,  turn
+     * it into a CrtAccessToken instance if present, and run it through the Authenticator.
+     */
+    @Override
+    public HeliosUser getValue(HttpContext c) {
+      final String authHeader = c.getRequest().getHeaderValue(HttpHeaders.AUTHORIZATION);
+
+      if (authHeader != null) {
+        final String[] tokenParts = authHeader.split(":");
+        if (tokenParts.length == 2) {
+          final String prefix = tokenParts[0];
+
+          if (prefix.equals("chap")) {
+            final String token = tokenParts[1];
+            try {
+              final Optional<HeliosUser> result =
+                  authenticator.authenticate(new CrtAccessToken(token));
+
+              if (result.isPresent()) {
+                return result.get();
+              }
+            } catch (AuthenticationException e) {
+              log.warn("Error authenticating credentials", e);
+              throw new WebApplicationException(Response.Status.INTERNAL_SERVER_ERROR);
+            }
+          }
+        }
+      }
+
+      // if we have fallen through here, because the header is not present or is not in the
+      // expected structure, return 401 if the auth token is required.
+      if (required) {
+        throw new WebApplicationException(
+            Response.status(Status.UNAUTHORIZED)
+                // TODO (mbrown): check if this header value is correct
```

Comment:
My interpretation of http://www.ietf.org/rfc/rfc2617.txt is that `realm` must always be included, but the text isn't trivial to parse so I could be wrong.